src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala

```diff
@@ -54,25 +55,32 @@ class PendingTransactionsMemPool[F[_]: Concurrent](
           case (addressHash, txs) =>
             transactionChainService
               .getLastAcceptedTransactionRef(addressHash)
-              .flatTap { last =>
-                logger.info(s"${Console.YELLOW}Last accepted: ${last} | Txs head: ${txs.headOption
-                  .map(_.transaction.lastTxRef)} hash=${txs.headOption.map(_.transaction.hash)}${Console.RESET}")
-              }
               .map { last =>
-                val ordinal = last.ordinal
-                val left = txs.filter(t => t.transaction.lastTxRef.ordinal >= ordinal || t.transaction.isTest) // TODO: get rid of >=
-                (last, left)
+                val consecutiveTransactions = takeConsecutiveTransactions(last, txs)
+                (last, consecutiveTransactions)
               }
-              .map {
-                case (last, t) =>
-                  if (t.headOption.map(_.transaction.lastTxRef).contains(last)) t
-                  else List.empty[TransactionCacheData]
+              .flatTap { case (last, consecTxs) =>
+                logger.info(s"${Console.YELLOW}Last accepted: ${last} | Txs head: ${consecTxs.headOption
+                  .map(_.transaction.lastTxRef)} hash=${consecTxs.headOption.map(_.transaction.hash)}${Console.RESET}")
               }
+              .map { case (_, consecTxs) => consecTxs }
         }
       }
       .map(
         _.sortBy(_.map(-_.transaction.fee.getOrElse(0L)).sum).flatten
       )
+
+  private def takeConsecutiveTransactions(lastAcceptedTxRef: LastTransactionRef, txs: List[TransactionCacheData]): List[TransactionCacheData] = {
+    @tailrec
+    def loop(acc: List[TransactionCacheData], prevTxRef: LastTransactionRef, txs: List[TransactionCacheData]): List[TransactionCacheData] =
+      txs.find(_.transaction.lastTxRef == prevTxRef) match {
+        case Some(tx) =>
+          loop(tx +: acc, LastTransactionRef(tx.transaction.hash, tx.transaction.ordinal), txs diff List(tx))
+        case None => acc.reverse //to preserve order of the chain
```

Comment:
so if you need to reverse the `acc` at the end, maybe just switch append/prepend in the loop?

Reply:
I guess it's less heavy with prepend and one reverse at the end as opposed to append for every element. Prepend is constant time, append is linear.